src/middlewared/middlewared/plugins/pool.py

```diff
@@ -2869,16 +2870,31 @@ class Config:
                 data
             )
 
+        kmip_config = await self.middleware.call('kmip.config')
+        if kmip_config['enabled'] and kmip_config['manage_zfs_keys']:
+            await self.middleware.call('kmip.sync_zfs_keys', [pk])
+
         return pk
 
     @private
-    def encrypted_roots_query_db(self, filters=None, decrypt=False):
-        return {
-            d['name']: d['encryption_key'] if not decrypt else self.middleware.call_sync(
-                'pwenc.decrypt', d['encryption_key']
-            )
-            for d in self.middleware.call_sync('datastore.query', self.dataset_store, filters or [])
-        }
+    def encrypted_roots_query_db(self, filters=None):
+        # We query database first - if we are able to find an encryption key, we assume it's the correct one.
+        # If we are unable to find the key in database, we see if we have it in memory with the KMIP server, if not,
+        # there are 2 ways this can go, we don't retrieve the key or the user can sync KMIP keys and we will have it
+        # with the KMIP service again through which we can retrieve them
+        datasets = self.middleware.call_sync('datastore.query', self.dataset_store, filters or [])
+        zfs_keys = self.middleware.call_sync('kmip.retrieve_zfs_keys')
+        keys = {}
+        for ds in datasets:
+            if ds['encryption_key']:
+                keys[ds['name']] = self.middleware.call_sync('pwenc.decrypt', ds['encryption_key'])
+            elif ds['name'] in zfs_keys:
+                keys[ds['name']] = zfs_keys[ds['name']]
+        return keys
+
+    @private
+    def dataset_datastore(self):
+        return self.dataset_store
```

Comment:
Why do you think this is necessary?

Reply:
I can retrieve dataset's table name via this method without hardcoding it everywhere

Reply:
And as the table has already been made and likely won't be changed ever, i think it's safe to use the name directly. I'll update this :p